Dyn 5497 markdown not loading

RobertGlobant20 opened the pull request:
### Purpose

Fixing bug about markdown files not loading for node packages documentation.
There was a code change that was preventing for loading md files for packages node documentation, so after this fix was done in a different pull request (https://github.com/DynamoDS/Dynamo/pull/13693), now it is loading the md file but the images were not loading due that WebView2 was using virtual folder fixed to fallback_docs folder then in this fix we are allowing to create the virtual folder dynamic depending if is a dynamo node or package node.
For some nodes of the Orchid package is working there is another bug with some specific nodes (like Math.Addition).

### Declarations

Check these if you believe they are true

- [X] The codebase is in a better state after this PR
- [X] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [X] The level of testing this PR includes is appropriate
- [X] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

Fixing bug about markdown files not loading for node packages documentation.


### Reviewers

@QilongTang @reddyashish @dnenov 

### FYIs

RobertGlobant20 commented:
Screenshot showing the documentation with images for a package node.
![image](https://user-images.githubusercontent.com/61755417/214168922-20296aeb-c7fe-460d-93e8-446eaa7f196d.png)

mjkkirschner commented on src/DocumentationBrowserViewExtension/DocumentationBrowserViewModel.cs:
```diff
@@ -132,6 +140,8 @@ public bool ShowBrowser
 
         internal UIElement DynamoView { get; set; }
 
+        internal NodeInfo CurrentNodeInfo { get; set; }
```
I don't think carrying around this state is ideal - at the minimum it should have some comments...

I'm not sure if it's possible, but it seems simpler to keep track of where this data is used if it's passed in an event or binding (or something similar) to the View.

Just a thought.

RobertGlobant20 replied:
@mjkkirschner removed the NodeInfo class and I just added a new internal property, check the next commit:
https://github.com/DynamoDS/Dynamo/pull/13703/commits/fdd337c2dd455d53ae51bbd88c17e8a5056ca389

mjkkirschner commented:
Can we write a test for this regression?

mjkkirschner commented:
what does this mean `For some nodes of the Orchid package is working there is another bug with some specific nodes (like Math.Addition).` Is it tracked and in jira?

RobertGlobant20 commented:
> Can we write a test for this regression?

@mjkkirschner please confirm that this a list of steps for the test expected for validating this functionality.
With this test we will be only validating that the md file expected will be loaded (no validation on images loaded).
- Open Dynamo
- Install a package with documentation (note that most of the packages currently doesn't have documentation even the Orchid package in production doesn't have documentation).
- Open the DocumentationBrowser
- Validate that the md in the Package/doc folder match with the one loaded in the DocumentationBrowser.

mjkkirschner commented:
> > Can we write a test for this regression?
> 
> @mjkkirschner please confirm that this a list of steps for the test expected for validating this functionality. With this test we will be only validating that the md file expected will be loaded (no validation on images loaded).
> 
> * Open Dynamo
> * Install a package with documentation (note that most of the packages currently doesn't have documentation even the Orchid package in production doesn't have documentation).
> * Open the DocumentationBrowser
> * Validate that the md in the Package/doc folder match with the one loaded in the DocumentationBrowser.

what about mocking most of that - for example, just set the view model data with some fake data that should come from a package, and then check that the virtual folder is setup correctly using that data - we just want to ensure it does not regress later.

filipeotero commented on src/DocumentationBrowserViewExtension/DocumentationBrowserViewModel.cs:
```diff
@@ -89,6 +90,22 @@ private set
+        /// <summary>
+        /// Package Name
+        /// </summary>
+        internal string Name
```
Maybe a more intuitive name would be PackageName... What do you think?

zeusongit replied:
Or graph name?

RobertGlobant20 replied:
@filipeotero yes, I think is a better idea to use PackageName, then updated in the next commit: https://github.com/DynamoDS/Dynamo/pull/13703/commits/0e3c928e47fc6f87b9d35cbe9b58b41d47318eda

RobertGlobant20 commented:
> > > Can we write a test for this regression?
> > 
> > 
> > @mjkkirschner please confirm that this a list of steps for the test expected for validating this functionality. With this test we will be only validating that the md file expected will be loaded (no validation on images loaded).
> > 
> > * Open Dynamo
> > * Install a package with documentation (note that most of the packages currently doesn't have documentation even the Orchid package in production doesn't have documentation).
> > * Open the DocumentationBrowser
> > * Validate that the md in the Package/doc folder match with the one loaded in the DocumentationBrowser.
> 
> what about mocking most of that - for example, just set the view model data with some fake data that should come from a package, open the docs browser, and then check that the virtual folder is setup correctly using that data - we just want to ensure it does not regress later.

@mjkkirschner I've added a unit test for validating that the WebView2 Virtual Directory is created in the right place (so images will be loaded correctly) when displaying documentation for package nodes.

When implementing the unit test I found that the call to await WebView2.EnsureCoreWebView2Async() was returning the execution to the unit test (instead of executing the next code), then I had to move the Virtual Folder code creation before EnsureCoreWebView2Async() call also I've added a property for VirtualDirectory so it can be checked in the unit test., please let me know is the test should be enough 
Thanks

Here is the commit with the changes:
https://github.com/DynamoDS/Dynamo/pull/13703/commits/0e3c928e47fc6f87b9d35cbe9b58b41d47318eda

RobertGlobant20 commented on src/DocumentationBrowserViewExtension/DocumentationBrowserView.xaml.cs:
```diff
@@ -130,6 +133,21 @@ protected virtual void Dispose(bool disposing)
+            {
+                if (viewModel.Link != null && !string.IsNullOrEmpty(viewModel.PackageName))
+                {
+                    VirtualFolderPath = HttpUtility.UrlDecode(viewModel.Link.AbsolutePath).Replace(viewModel.PackageName + ".md", "");
```
thanks @pinzart90 I've done the change in the next commit: https://github.com/DynamoDS/Dynamo/pull/13703/commits/ad5209f44adda39a2cd3e0f619a5991c479e3c5a

mjkkirschner commented on src/DocumentationBrowserViewExtension/DocumentationBrowserView.xaml.cs:
```diff
@@ -130,6 +133,21 @@ protected virtual void Dispose(bool disposing)
+                else
+                    VirtualFolderPath = FallbackDirectoryName;
+            }
+            catch (Exception ex)
```
what exception do you think will be caught here?

RobertGlobant20 replied:
@mjkkirschner I found a case that `viewModel.Link = "{}"` so it was crashing when trying the next line `viewModel.Link.AbsolutePath` so I added this try catch because I didn't found any other way to validate the `viewModel.Link` content.
With the fix added related to DSVarArgFunction now I expect that viewModel.Link always will contain something valid but just in case I think we should keep this code or if you have any other idea in how to validate  `viewModel.Link = "{}"` please let me know (Link is Uri type, Uri.IsWellFormedUriString returns true).

mjkkirschner replied:
I think it's a good practice to log something including the exception if possible to the dynamo console - if some unanticipated exception starts showing up here atleast we'll know.

RobertGlobant20 replied:
@mjkkirschner I've added some code for logging the exception in the dynamo console, see the next commit: https://github.com/DynamoDS/Dynamo/pull/13703/commits/5980d151914322d4a1eb6d6965f5e2b116cfc9a4

mjkkirschner commented on src/DocumentationBrowserViewExtension/DocumentationBrowserViewModel.cs:
```diff
@@ -87,7 +88,23 @@ private set
+        /// <summary>
+        /// Package Name
+        /// </summary>
+        internal string PackageName
```
it still feels pretty odd to me that the DocumentationBrowserViewModel has a single piece of state like this that it has to hold on to.

If I am understanding correctly, can we at least name it something like:
`CurrentPackageName` or something like that and add a comment describing when it will be null and when it will be valid?

RobertGlobant20 replied:
@mjkkirschner notice that the dependency on PackageName was added in the next PR:
https://github.com/DynamoDS/Dynamo/pull/13414
Renamed to CurrentPackageName in the next commit: https://github.com/DynamoDS/Dynamo/pull/13703/commits/f00db5b919450b0ccdafac4c9f90a8827b5d80c1
![image](https://user-images.githubusercontent.com/61755417/216129140-c5d1f11d-1376-4aad-b901-60ce3a393f5c.png)

mjkkirschner commented on src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs:
```diff
@@ -2120,9 +2120,9 @@ internal AssemblyName GetNameOfAssemblyReferencedByNode(NodeModel node)
             AssemblyName assemblyName = null;
             // Get zerotouch assembly
-            if (node is DSFunction)
+            if (node is DSFunction || node is DSVarArgFunction)
```
I am incorrect, `DSVarArgFunction` inherits from `DSFunctionBase` not `DSFunction` - apologies for the confusion.

mjkkirschner commented on src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs:
```diff
@@ -2120,9 +2120,9 @@ internal AssemblyName GetNameOfAssemblyReferencedByNode(NodeModel node)
+            if (node is DSFunction || node is DSVarArgFunction)
             {
-                var descriptor = (node as DSFunction).Controller.Definition;
+                var descriptor = node is DSFunction function ? function.Controller.Definition :  (node as DSVarArgFunction).Controller.Definition;
```
I think you can get rid of all these cases and the if statement above - just check if the type is `DSFunctionBase` - it has a `Controller.Definition` property you can access and you don't need to check any types, it's safer an faster - unless you need special behavior for each type.

RobertGlobant20 replied:
Checking if is DSFunctionBase type I've simplified the code in the next way:
![image](https://user-images.githubusercontent.com/61755417/216139960-fe6acdeb-6edc-4fca-a803-84c3010424fd.png)
Please let me know if it sounds good to you. Thanks
commit: https://github.com/DynamoDS/Dynamo/pull/13703/commits/f00db5b919450b0ccdafac4c9f90a8827b5d80c1

mjkkirschner commented on test/DynamoCoreWpfTests/ViewExtensions/DocumentationBrowserViewExtensionTests.cs:
```diff
@@ -103,6 +103,48 @@ public void ShowingStartPageHidesBrowser()
+
+            docBrowserviewExtension.HandleRequestOpenDocumentationLink(nodeAnnotationEventArgs);
+            var htmlContent = GetSidebarDocsBrowserContents();
+            htmlContent = htmlContent.Replace(@"%5C", "/");
```
please add a comment

RobertGlobant20 replied:
Added comment, please check the next commit: https://github.com/DynamoDS/Dynamo/pull/13703/commits/f00db5b919450b0ccdafac4c9f90a8827b5d80c1

mjkkirschner reviewed (changes requested):
please see inline comments

mjkkirschner commented:
thanks @RobertGlobant20 !

zeusongit commented:
Thanks for addressing the review comments @RobertGlobant20 I will merge the PR as soon as the self serve job completes

RobertGlobant20 commented:
> Thanks for addressing the review comments @RobertGlobant20 I will merge the PR as soon as the self serve job completes
@zeusongit 
Just to comment that I keep receiving this emails after I push any commit in this PR (but all the checks seems that are passing):
![image](https://user-images.githubusercontent.com/61755417/216479182-66141f09-56a1-4a3a-9bc3-1d88fa9b587d.png)

mjkkirschner commented:
@RobertGlobant20 seems there is a conflicted file.
